elf_load.c

```diff
@@ -311,13 +312,132 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
 }
 
 #if CONFIG_LIBVFSCORE
+#if CONFIG_LIBPOSIX_MMAP
+static int elf_load_fdphdr(struct elf_prog *elf_prog, GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart, vaend;
+	uint64_t delta_p_offset;
+	static void *addr;
+	static int flags;
+
+	/* If we are mmapping a fresh ELF, restore address and flags */
+	if (!phdr->p_paddr) {
+		addr = NULL;
+		flags = MAP_PRIVATE;
+	}
+
+	/* PT_LOAD's paddr may be misaligned, so keep in mind the
+	 * offset from what should have been the aligned paddr to
+	 * what we ended up with.
+	 */
+	delta_p_offset = phdr->p_paddr - PAGE_ALIGN_DOWN(phdr->p_paddr);
+
+	/* If this is not the first mmap'd ELF segment, then keep
+	 * contiguously mapping from where we left off.
+	 */
+	if (elf_prog->vabase)
+		addr = (void *)PAGE_ALIGN_DOWN((phdr->p_paddr +
+					       (uintptr_t)elf_prog->vabase));
+
+	uk_pr_debug("%s: Memory mapping 0x%"PRIx64" - 0x%"PRIx64" to 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)phdr->p_offset,
+		    (uint64_t)phdr->p_filesz + delta_p_offset,
+		    (uint64_t)addr,
+		    (uint64_t)addr + (uint64_t)phdr->p_filesz +
+		    delta_p_offset);
+
+	/* mmap with all flags. If protections are enabled, these will
+	 * be manually re-adjusted later.
+	 */
+	vastart = (uintptr_t)mmap(addr, phdr->p_filesz + delta_p_offset,
```

Comment:
Will this map data outside current segment? Should the un-aligned regions be memset to 0?

Reply:
> Will this map data outside current segment?

Good point :+1:. But is it possible for a PT_LOAD to have an alignment smaller than page-size?

> Should the un-aligned regions be memset to 0?

I am not sure I understand :thinking:. Allow me to give you an example of what I do with misaligned distances between ELF segments. Suppose you have the following segments (these are from NodeJs v18.15.0 btw :smiley:):
```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr           FileSiz            MemSiz         Flags  Align
  LOAD           0x0000000001a00000 0x0000000001a00000 0x0000000001a00000 0x0000000000000231 0x0000000000000231 R E 0x200000
  LOAD           0x0000000001a01000 0x0000000001a01000 0x0000000001a01000 0x0000000000d056f0 0x0000000000d056f0 R   0x200000
  LOAD           0x0000000002706af0 0x0000000002707af0 0x0000000002707af0 0x0000000000083d70 0x00000000000a0810 RW  0x200000
```

What will happen:
First PT_LOAD:
- I mmap `0x1a00000 -> 0x1a00231` => `0x1a00000 -> 0x1a01000` ends up being mmap'd
- since `0x1a01000 - 0x1a00231 == 0xdcf` which is less than `PAGE_SIZE`, I memset that remainder to `0`

Second PT_LOAD:
- I mmap `0x1a01000 -> 0x27066f0` => `0x1a01000 -> 0x2707000` ends up being mmap'd
- since `0x2707000 - 0x27066f0 == 0x910` which is less than `PAGE_SIZE`, I memset that remainder to `0`

Third PT_LOAD:
- I mmap `0x2707000` (this is PAGE_ALIGN_DOWN(0x2707af0)) ` -> 0x278b860` (0x2707000 + 0x83d70 + 0xaf0 (delta_p_offset)) => `0x2707000 -> 0x278c000` ends up being mmap'd
- since `0x278c000 - 0x278b860 == 0x7a0` which is less than `PAGE_SIZE`, I memset that remainder to `0`
- `MemSiz` (`0xa0810`) is bigger than `FileSiz` (`0x83d70`) => We mmap `0x278c000 -> 0x27a8300` as anonymous pages, this way we avoid having to memset the entire `.bss`, which can be quite large. So this ends up mmap-ing `0x278c000 -> 0x27a9000`.

Could you give me an example where the current code would not go as expected? :thinking: I am sorry, I am not entirely sure I understand the pitfall you are suggesting.

Reply:
You are right, sorry for my mistake. I have found the ABI document, and it says "executable and shared object files must have segments whose file offsets and virtual addresses are congruent modulo the page size".

Reply:
> sorry for my mistake

Don't apologize. I encourage you to question as many things as possible in this PR as well as others, let's make sure the code looks clean and works correctly from everyone's perspective :smiley:. Thank you for investing the time into looking things up in the ABI :muscle:, I appreciate it. I also just now remember having read these exact words in the past but I also forgot (you just can't make this up: "congruent modulo" :sweat_smile:, sounds super :brain:).

Reply:
Nope, I think it is correct, `0x0000000002706af0` and `0x0000000002707af0` give the same remainder on division with page size, right? This is taken care by `deltap_p_offset`.